core/drivers/crypto/crypto_api/hash/hash.c

```diff
@@ -0,0 +1,310 @@
+
+		/*
+		 * Verify that the HASH HW implements this algorithm
+		 * else return NULL pointer to call the Hash SW Library
+		 */
+		if (hash) {
+			if (hash->max_hash < hash_id)
+				hash = NULL;
+		}
+
+		/* Verify that the HASH HW implements this algorithm */
+		if (id)
+			*id = hash_id;
+	}
+
+	CRYPTO_TRACE("Check Hash algo %d ret 0x%"PRIxPTR"",
+		algo_id, (uintptr_t)hash);
+
+	return hash;
+}
+
+/*
+ * Free the Software Hashing Context function of the algorithm
+ *
+ * Inputs:
+ * ctx    Reference the API context
+ *
+ */
+static void hash_free_ctx(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if (hash) {
+		if (hash->op) {
+			if (hash->op->free_ctx)
+				hash->op->free_ctx(hash->ctx);
+		}
+		free(hash);
+	}
+}
+
+/*
+ * Copy Software Hashing Context
+ *
+ * Inputs:
+ * src_ctx  Reference the API context source
+ *
+ * Outputs:
+ * dst_ctx  Reference the API context destination
+ *
+ */
+static void hash_copy_state(struct crypto_hash_ctx *dst_ctx,
+		struct crypto_hash_ctx *src_ctx)
+{
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
+
+	if ((!dst_ctx) || (!src_ctx))
+		return;
+
+	if (hash_src->op) {
+		if (hash_src->op->cpy_state)
+			hash_src->op->cpy_state(hash_dst->ctx, hash_src->ctx);
+	}
+}
+
+/*
+ * Initialization of the Hash operation
+ *
+ * Inputs:
+ * ctx   Reference the API context
+ *
+ * Returns:
+ * TEE_SUCCESS               Success
+ * TEE_ERROR_BAD_PARAMETERS  Bad parameters
+ * TEE_ERROR_NOT_IMPLEMENTED Algorithm is not implemented
+ */
+static TEE_Result hash_init(struct crypto_hash_ctx *ctx)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if (!hash)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (hash->op) {
+		if (hash->op->init)
+			ret = hash->op->init(hash->ctx);
+	}
+
+	CRYPTO_TRACE("hash ret 0x%"PRIX32"", ret);
+
+	return ret;
+}
+
+/*
+ * Update the Hash operation
+ *
+ * Inputs:
+ * ctx   Reference the API context
+ * data  Data to hash
+ * len   Data length
+ *
+ * Returns:
+ * TEE_SUCCESS               Success
+ * TEE_ERROR_GENERIC         Other Error
+ * TEE_ERROR_BAD_PARAMETERS  Bad parameters
+ * TEE_ERROR_NOT_IMPLEMENTED Algorithm is not implemented
+ */
+static TEE_Result hash_update(struct crypto_hash_ctx *ctx,
+		const uint8_t *data, size_t len)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if ((!hash) || ((!data) && (len != 0)))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (hash->op) {
+		if (hash->op->update)
+			ret = hash->op->update(hash->ctx, data, len);
+	}
+
+	CRYPTO_TRACE("hash ret 0x%"PRIX32"", ret);
+	return ret;
+}
+
+/*
+ * Finalize the Hash operation
+ *
+ * Inputs:
+ * ctx   Reference the API context
+ * len   Digest buffer length
+ *
+ * Outputs:
+ * digest  Hash digest buffer
+ *
+ * Returns:
```

Comment:
minor: no real need to detail the return values here.

Reply:
Could you please say me what are the function header rules you apply?

Reply:
As you can see browsing optee_os source files there is no unique way for function description.